File: src/test/java/com/automation/framework/driver/DriverFactory.java
package com.automation.framework.driver;

import com.automation.framework.utils.ConfigReader;
import io.github.bonigarcia.wdm.WebDriverManager;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxOptions;

import java.time.Duration;

public final class DriverFactory {

  private static final Logger log = LogManager.getLogger(DriverFactory.class);

  private DriverFactory() {}

  public static void initDriver() {
    // ConfigReader already resolves System properties / env vars and ignores blank overrides.
    String browser = ConfigReader.get("browser").trim().toLowerCase();
    boolean headless = ConfigReader.getBoolean("headless", true);

    WebDriver driver;
    switch (browser) {
      case "firefox" -> {
        WebDriverManager.firefoxdriver().setup();
        FirefoxOptions options = new FirefoxOptions();
        if (headless) {
          options.addArguments("-headless");
        }
        driver = new FirefoxDriver(options);
      }
      case "chrome" -> {
        WebDriverManager.chromedriver().setup();
        ChromeOptions options = new ChromeOptions();
        if (headless) {
          options.addArguments("--headless=new");
        }
        options.addArguments("--no-sandbox");
        options.addArguments("--disable-dev-shm-usage");
        driver = new ChromeDriver(options);
      }
      default -> throw new IllegalArgumentException("Unsupported browser: " + browser);
    }

    int implicitWait = ConfigReader.getInt("implicitWaitSeconds", 2);
    driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(implicitWait));
    driver.manage().window().maximize();

    DriverManager.setDriver(driver);
    log.info("Initialized WebDriver. browser={}, headless={}", browser, headless);
  }

  public static void quitDriver() {
    try {
      WebDriver driver = DriverManager.getDriver();
      driver.quit();
    } catch (Exception ignored) {
      // ignore
    } finally {
      DriverManager.unload();
    }
  }
}

File: python/framework/driver/driver_factory.py
import logging

from selenium import webdriver
from selenium.webdriver.chrome.options import Options as ChromeOptions
from selenium.webdriver.chrome.service import Service as ChromeService
from selenium.webdriver.firefox.options import Options as FirefoxOptions
from selenium.webdriver.firefox.service import Service as FirefoxService
from webdriver_manager.chrome import ChromeDriverManager
from webdriver_manager.firefox import GeckoDriverManager

from framework.driver.driver_manager import DriverManager
from framework.utils.config_reader import get, get_bool, get_int

log = logging.getLogger(__name__)


def init_driver():
    browser = get("browser").strip().lower()
    headless = get_bool("headless", False)

    if browser == "firefox":
        options = FirefoxOptions()
        if headless:
            options.add_argument("-headless")
        service = FirefoxService(GeckoDriverManager().install())
        driver = webdriver.Firefox(service=service, options=options)
    elif browser == "chrome":
        options = ChromeOptions()
        if headless:
            options.add_argument("--headless=new")
        options.add_argument("--no-sandbox")
        options.add_argument("--disable-dev-shm-usage")
        service = ChromeService(ChromeDriverManager().install())
        driver = webdriver.Chrome(service=service, options=options)
    else:
        raise ValueError(f"Unsupported browser: {browser}")

    implicit_wait = get_int("implicitWaitSeconds", 2)
    driver.implicitly_wait(implicit_wait)

    try:
        driver.maximize_window()
    except Exception:
        driver.set_window_size(1920, 1080)

    DriverManager.set_driver(driver)
    log.info("Initialized WebDriver. browser=%s, headless=%s", browser, headless)
    return driver


def quit_driver():
    try:
        driver = DriverManager.get_driver()
        driver.quit()
    except Exception:
        pass
    finally:
        DriverManager.unload()
